### ISSUE TITLE
Internal Error when using eval function with Basic Iterative Attack Example on CIFAR10 dataset

### ISSUE THREAD

**Issue by GuanxiongLiu** [cleverhans-lab/cleverhans#874]
***The issue tracker should only be used to report bugs or feature requests. If you are looking for support from other library users, please ask a question on [StackOverflow](https://stackoverflow.com/questions/tagged/cleverhans).***

**Describe the bug**
I get an internal issue when I use the eval function with BIM attack examples on CIFAR10 dataset.

**To Reproduce**
1. The simple version code is provided.
[reproduce-code.zip](https://github.com/tensorflow/cleverhans/files/2535354/reproduce-code.zip)
2. Run the reproduction code.

**Expected behavior**
The code will show internal error on evaluating BIM accuracy. The error message points to while loop.

**Screenshots**
![capture](https://user-images.githubusercontent.com/6549940/47803017-ea499300-dd07-11e8-996b-f3dcebfd6ddf.PNG)

**System configuration**
 - OS Ubuntu 17.10
 - Python version 3.6
 - TensorFlow version 1.9

**Comment by npapernot** [comment]
Looks like this is related to a recent issue encountered by @carlini : https://github.com/tensorflow/tensorflow/issues/23166

@carlini could you comment if that's the case?

**Comment by carlini** [comment]
Yep. This is exactly why I raised that issue. I have a short reproduction over there.

**Comment by rameshKrSah** [comment]
The same issue occurred when using Basic Iterative attack on a neural network with dropout layers. Reading further I found that setting Keras backend learning phase to 0 (i.e. test) solved the problem.

I don't know why setting the learning phase to 0 solved the problem but this may be a temporary fix for now.

**Comment by jpgard** [comment]
Just following up because it is not entirely clear from the discussion above -- does this mean `BasicIterativeMethod` is broken? If not, can you clarify under what circumstances `BasicIterativeMethod` should work, and when it won't work?

Setting the learning phase did not resolve problems for me. Specifically, the issue I am experiencing is that model compilation hangs permanently when using `BasicIterativeMethod` or `ProjectedGradientDescent` in the adversarial acc metric (`model.compile()` never finishes) - it seems like that might be related to the same underlying issue.

**Comment by carlini** [comment]
This is a bug in tensorflow. There's not much we can do other than work around it, or use something other than tensorflow.

This bug always results in an ugly stackdump. If code hangs it's not a result of this bug.

**Comment by jpgard** [comment]
Got it, thank you.

**Comment by jpgard** [comment]
To clarify, does this mean `BasicIterativeMethod` should never be expected to work? Under what circumstances does it fail?

**Comment by carlini** [comment]
No, it does work sometimes. We don't really understand when it does and doesn't. It has to do with keras and batch norm it seems, at least, but might fail under other situations too.

**Comment by jpgard** [comment]
Thanks. 

To add to that, I can reproduce the same issue -- model hangs forever at compile stage -- even when using the `mnist_tutorial_keras.py`. Simply swap the `FastGradientMethod` instance with an instance of `ProjectedGradientDescent` in that tutorial.

**Comment by jonasguan** [comment]
Closing as we discontinue support for TF1. Too bad they never fixed it :(
